drivers/i2c/i2c_rpi_pico_pio.c

```diff
@@ -0,0 +1,567 @@
+	pio_i2c_rx_enable(pio, sm, false);
+	pio_i2c_put16(pio, sm, (addr << 2) | 1u);
+	while (len && !pio_i2c_check_error(pio, sm)) {
+		if (!pio_sm_is_tx_fifo_full(pio, sm)) {
+			--len;
+			pio_i2c_put_or_err(pio, sm,
+					   (*txbuf++ << PIO_I2C_DATA_LSB) |
+						   ((len == 0) << PIO_I2C_FINAL_LSB) | 1u);
+		} else {
+			k_yield();
+		}
+	}
+	if (flags & I2C_MSG_STOP) {
+		pio_i2c_stop(pio, sm);
+	}
+	pio_i2c_wait_idle(pio, sm);
+	if (pio_i2c_check_error(pio, sm)) {
+		err = -1;
+		pio_i2c_resume_after_error(pio, sm);
+		pio_i2c_stop(pio, sm);
+	}
+	return err;
+}
+
+static int pio_i2c_read_blocking(PIO pio, uint sm, uint8_t addr, uint8_t *rxbuf, uint len,
+				 uint8_t flags)
+{
+	int err = 0;
+
+	if (flags & I2C_MSG_RESTART) {
+		pio_i2c_repstart(pio, sm);
+	} else {
+		pio_i2c_start(pio, sm);
+	}
+	pio_i2c_rx_enable(pio, sm, true);
+	while (!pio_sm_is_rx_fifo_empty(pio, sm)) {
+		(void)pio_i2c_get(pio, sm);
+	}
+	pio_i2c_put16(pio, sm, (addr << 2) | 3u);
+	uint32_t tx_remain = len; /* Need to stuff 0xff bytes in to get clocks */
+
+	bool first = true;
+
+	while ((tx_remain || len) && !pio_i2c_check_error(pio, sm)) {
+		if (tx_remain && !pio_sm_is_tx_fifo_full(pio, sm)) {
+			--tx_remain;
+			pio_i2c_put16(pio, sm,
+				      (0xffu << 1) | (tx_remain ? 0
+								: (1u << PIO_I2C_FINAL_LSB) |
+									  (1u << PIO_I2C_NAK_LSB)));
+		}
+		if (!pio_sm_is_rx_fifo_empty(pio, sm)) {
+			if (first) {
+				/* Ignore returned address byte */
+				(void)pio_i2c_get(pio, sm);
+				first = false;
+			} else {
+				--len;
+				*rxbuf++ = pio_i2c_get(pio, sm);
+			}
+		} else {
+			k_yield();
+		}
+	}
+	if (flags & I2C_MSG_STOP) {
+		pio_i2c_stop(pio, sm);
+	}
+	pio_i2c_wait_idle(pio, sm);
+	if (pio_i2c_check_error(pio, sm)) {
+		err = -1;
+		pio_i2c_resume_after_error(pio, sm);
+		pio_i2c_stop(pio, sm);
+	}
+	return err;
+}
+
+/**@}*/
+
+static int i2c_pio_configure(const struct device *dev, uint32_t dev_config)
+{
+	struct i2c_pio_data *data = dev->data;
+
+	if (I2C_SPEED_GET(dev_config) != I2C_SPEED_DT) {
+		return -ENOTSUP;
+	}
+
+	data->config = dev_config;
+	return 0;
+}
+
+static int i2c_pio_get_config(const struct device *dev, uint32_t *config)
+{
+	struct i2c_pio_data *data = dev->data;
+
+	return data->config;
+}
+
+static int i2c_pio_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
+			    uint16_t addr)
+{
+	const struct i2c_pio_config *cfg = dev->config;
+	struct i2c_pio_data *data = dev->data;
+
+	PIO pio = pio_rpi_pico_get_pio(cfg->piodev);
+	uint sm = cfg->sm;
+	int rc;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	for (int i = 0; i < num_msgs; i++) {
+		if (msgs[i].flags & I2C_MSG_READ) {
+			rc = pio_i2c_read_blocking(pio, sm, addr, msgs[i].buf, msgs[i].len,
+						   msgs[i].flags);
+		} else {
+			rc = pio_i2c_write_blocking(pio, sm, addr, msgs[i].buf, msgs[i].len,
+						    msgs[i].flags);
+		}
+		if (rc) {
+			rc = -EIO;
+			goto done;
+		}
+	}
+
+	rc = 0;
+done:
+	k_mutex_unlock(&data->lock);
+	return 0;
```

Comment:
```suggestion
	return rc;
```
else a NACK does not lead to an error (tested at a rp2040 with a wrong slave address configured )